continued: Remove Module Base Address and Load Bias from FunctionInfo

antonrohr opened the pull request:
This continues PR #1315 

Everything should now work as expected, aka the context menus in the different views should be correct and show options like "hook" only if Orbit is connected and the correct process is selected. 

I marked this as draft, because I can't help but think this is currently more complicated than it needs to be. The commit that moves "module_map" out of capture data seemed necessary, because capture deserialization is done off the main thread and the "module_map" is stored in DataManager which can only be accessed from the main thread. 
I would propose to change the deserialization instead and still keep a copy of the module_map in CaptureData as it was done before. I think this would simplify some function calls, but I am unsure if its the best way to go forward. wdyt? Maybe one of you has a good idea on a conceptual level.

florian-kuebler commented on OrbitGl/CallStackDataView.cpp:
```diff
@@ -191,10 +190,6 @@ CallStackDataView::CallStackDataViewFrame CallStackDataView::GetFrameFromRow(int
 
 CallStackDataView::CallStackDataViewFrame CallStackDataView::GetFrameFromIndex(
     int index_in_callstack) {
-  if (index_in_callstack >= static_cast<int>(callstack_.GetFramesCount())) {
```
I guess you are removing this, as this does never happen. Could you then add a CHECK for now?

antonrohr replied:
done

florian-kuebler commented on OrbitGl/SamplingReportDataView.cpp:
```diff
@@ -49,6 +49,8 @@ std::string SamplingReportDataView::GetValue(int row, int column) {
     case kColumnInclusive:
       return absl::StrFormat("%.2f", func.inclusive);
     case kColumnModuleName:
+      // LOG("%s", func.module_path);
```
Can you remove those?

antonrohr replied:
done

florian-kuebler commented on OrbitGl/SamplingReportDataView.cpp:
```diff
@@ -159,22 +161,25 @@ absl::flat_hash_set<const FunctionInfo*> SamplingReportDataView::GetFunctionsFro
+    if (!result) {
+      ERROR("result %s", result.error().message());
+    }
+    if (result) {
```
why not just else?

antonrohr replied:
done

florian-kuebler commented on OrbitGl/LiveFunctionsDataView.cpp:
```diff
@@ -163,21 +163,23 @@ std::vector<std::string> LiveFunctionsDataView::GetContextMenu(
+      // Is that function actually inside a module of the process (i.e. can we disassemble)?
+      const FunctionInfo* actual_function =
+          capture_data.FindFunctionByAddress(absolute_address, module_map, false);
+      const bool function_exists = actual_function != nullptr;
```
I guess, this can now be a check? What happens if you have a process running for long time. Start orbit in a connected state, and load a capture of that process (which is still running). Can you hook then? Currently, we would need to reload the symbols, right?

antonrohr replied:
Now: If the same process is still running (determined by isCaptureConnected) then you can select/unselect and disassemble

florian-kuebler commented on OrbitGl/DataManager.h:
```diff
@@ -39,6 +41,7 @@ class DataManager final {
   void set_selected_process(int32_t pid);
 
   [[nodiscard]] const ProcessData* GetProcessByPid(int32_t process_id) const;
+  [[nodiscard]] ProcessData* GetMutableProcessByPid(int32_t process_id);
```
Where do you use that? In general, I think we should keep those accessors minimal. They basically allow us to completely shortcut the thread-safety measurements of DataManager (here, by passing that pointer to another thread).

antonrohr replied:
Yes, I agree, but it is necessary for one specific thing: Updating the loaded module list of a process. The tradeoff here is between having this function and writing an extra function that updates this list inside of DataManager. I decided for this tradeoff, but I am actually not sure anymore. Let me know what you think.

florian-kuebler replied:
I would argue for the function in data manager. But have no strong opinion on it. So I think it is fine for now. Thanks for the explanation.

florian-kuebler reviewed (commented):
Thanks for continuing that work. LGTM. I left some comments and questions.

dimitry- commented on OrbitCaptureClient/CaptureClient.cpp:
```diff
@@ -69,7 +63,7 @@ ErrorMessageOr<void> CaptureClient::StartCapture(
 
 void CaptureClient::Capture(ProcessData&& process,
-                            absl::flat_hash_map<std::string, ModuleData*>&& module_map,
+                            absl::flat_hash_map<std::string, ModuleData*> module_map,
```
should this be a const ref?

florian-kuebler commented on OrbitGl/App.cpp:
```diff
@@ -155,25 +154,26 @@ void OrbitApp::OnCaptureStarted(ProcessData&& process,
-      [this, sampling_profiler = std::move(sampling_profiler)]() mutable {
-        RefreshCaptureView();
+  main_thread_executor_->Schedule([this]() {
+    SamplingProfiler sampling_profiler(
```
Can you trie to move this one line above again to do this on the capture thread?

antonrohr replied:
done

antonrohr commented:
@dimitry- & @florian-kuebler 

This PR is finally in a ready for review state. I cleaned up a lot and squashed (fixup-ed) a few commits into one, because before there were changes introduced that I removed again. 

The commit messages should be a good description what is happening. Please look there. 

I tested a bunch (including some weird edge cases) and I think this is in a good shape, any further testing would be great of course.

florian-kuebler reviewed (approved):
Thanks Anton for those changes! I really like the new ModuleManager.
LGTM!

dimitry- commented on OrbitClientProtos/capture_data.proto:
```diff
@@ -106,6 +122,8 @@ message CaptureInfo {
   repeated TracepointInfo tracepoint_infos = 10;
   repeated TracepointEventInfo tracepoint_event_infos = 11;
+  ProcessInfo process = 13;
+  repeated ModuleInfo modules = 14;
```
ok, looks like moving these dependencies and CaptureInfo out of proto is much larger task, lets leave it as is.

dimitry- commented on OrbitGl/App.cpp:
```diff
@@ -785,6 +781,17 @@ bool OrbitApp::SelectProcess(const std::string& process) {
+  CHECK(capture_process != nullptr);
+
+  return selected_process->pid() == capture_process->pid() &&
+         selected_process->full_path() == capture_process->full_path();
```
TODO: check buildid?

antonrohr replied:
I put a TODO and it's a good idea to think about the details of what that would actually mean. Are build-ids required and  so on.

dimitry- reviewed (approved):
LGTM, thanks!
